runtime/parachains/src/paras_inherent.rs

```diff
@@ -290,239 +459,1658 @@ pub mod pallet {
+	});
+	ensure2!(backed_candidates.len() == n, Error::<T>::CandidateConcludedInvalid, EARLY_RETURN);
+
+	// Assure the backed candidate's `ParaId`'s core is free.
+	// This holds under the assumption that `Scheduler::schedule` is called _before_.
+	// Also checks the candidate references the correct relay parent.
+
+	let scheduled_paras_set = scheduled
+		.into_iter()
+		.map(|core_assignment| core_assignment.para_id)
+		.collect::<BTreeSet<_>>();
+	backed_candidates.retain(|backed_candidate| {
+		let desc = backed_candidate.descriptor();
+		desc.relay_parent == relay_parent && scheduled_paras_set.contains(&desc.para_id)
+	});
+	ensure2!(backed_candidates.len() == n, Error::<T>::CandidateConcludedInvalid, EARLY_RETURN);
+
+	// Limit weight, to avoid overweight block.
+	Ok(backed_candidates)
+}
+
+fn limit_disputes<T: Config>(disputes: &mut MultiDisputeStatementSet, entropy: [u8; 32]) -> Weight {
+	let mut remaining_weight = <T as frame_system::Config>::BlockWeights::get().max_block;
+	let disputes_weight = dispute_statements_weight::<T>(&disputes);
+	if disputes_weight > remaining_weight {
+		// Sort the dispute statements according to the following prioritization:
+		//  1. Prioritize local disputes over remote disputes.
+		//  2. Prioritize older disputes over newer disputes.
+		disputes.sort_by(|a, b| {
+			let a_local_block = T::DisputesHandler::included_state(a.session, a.candidate_hash);
+			let b_local_block = T::DisputesHandler::included_state(b.session, b.candidate_hash);
+			match (a_local_block, b_local_block) {
+				// Prioritize local disputes over remote disputes.
+				(None, Some(_)) => Ordering::Greater,
+				(Some(_), None) => Ordering::Less,
+				// For local disputes, prioritize those that occur at an earlier height.
+				(Some(a_height), Some(b_height)) => a_height.cmp(&b_height),
+				// Prioritize earlier remote disputes using session as rough proxy.
+				(None, None) => a.session.cmp(&b.session),
 			}
+		});
 
-			true
+		// Since the disputes array is sorted, we may use binary search to find the beginning of
+		// remote disputes
+		let idx = disputes
+			.binary_search_by(|probe| {
+				if T::DisputesHandler::included_state(probe.session, probe.candidate_hash).is_some()
+				{
+					Ordering::Greater
+				} else {
+					Ordering::Less
+				}
+			})
+			// The above predicate will never find an item and therefore we are guaranteed to obtain
+			// an error, which we can safely unwrap. QED.
+			.unwrap_err();
+
+		// Due to the binary search predicate above, the index computed will constitute the beginning
+		// of the remote disputes sub-array
+		let remote_disputes = disputes.split_off(idx);
+
+		// Select disputes in-order until the remaining weight is attained
+		disputes.retain(|d| {
+			let dispute_weight = <<T as Config>::WeightInfo as WeightInfo>::enter_variable_disputes(
+				d.statements.len() as u32,
+			);
+			if remaining_weight > dispute_weight {
```

Comment:
```suggestion
			if remaining_weight >= dispute_weight {
```

Reply:
I would prefer this to use `checked_sub()`, but I am not sure this makes it any more legible, but it would prevent us from having above issues.